### ISSUE TITLE
Validation metrics in MMSegWandbHook

### ISSUE THREAD

**Issue by kimc19** [open-mmlab/mmsegmentation#2051]
Hi,

Thanks for all your work.

I'm trying to visualize the validation metrics in Weights and Biases using the MMSegWandbHook.
I use the following settings:

```
log_config = dict(
    interval=1000,
    hooks=[
        dict(type='TextLoggerHook', by_epoch=False),
        dict(
            type='MMSegWandbHook',
            init_kwargs=dict(
                entity='kimc19',
                project='STDC_Pruebas',
                name='stdc_base'
            ),
            log_checkpoint=True,
            log_checkpoint_metadata=True,
            num_eval_images=100)
    ])
```

With this settings I'm able to visualize train metrics but not validation metrics.

Here is an example of what I the results I got in Weights and biases
![example wandb](https://user-images.githubusercontent.com/79665540/189512464-b4fefadd-0fe7-409f-8502-2cb328805836.png)

What could I try to visualize validation metrics in weights and biases?

**Comment by morganmcg1** [comment]
Hey @kimc19 , I'm from Weights & Biases, I'll ask one of the guys here to look into this, thanks for reporting!

**Comment by SebasGarcia08** [comment]
Hi, I'm looking forward to this as well. It seems like only the training images are logged because the EvalHook is added only on train hook events, but not on the validation ones. In my case, I'm specifying the hook like this:
```
log_config = dict(
    interval=1,
    hooks=[
        dict(type='TextLoggerHook', by_epoch=False),
        dict(
            type='MMSegWandbHook',
            by_epoch=False,
            interval=1,
            init_kwargs=dict(
                entity='sebastian-garcia-acosta',
                project='marsemseg',
                name='fcn_unet_s5-d16_4x4_512x1024_160k_cityscapes'),
            log_checkpoint=True,
            log_checkpoint_metadata=True,
            num_eval_images=10)
    ])
```
 And this is the execution order of the hooks:
 
![image](https://user-images.githubusercontent.com/46965003/189512146-45c21cab-a944-4d67-98f4-a60c08787ad2.png)

I'm looking forward in order to solve this bug.

Thank you very much.

**Comment by SebasGarcia08** [comment]
Quick update! **I managed to solved it by specifying`with_step=False` in the  MMSegWandbHook spec**. This is how my log_config looks now:

```
log_config = dict(
    interval=1,
    hooks=[
        dict(type='TextLoggerHook', by_epoch=False),
        dict(
            type='MMSegWandbHook',
            by_epoch=False,
            interval=1,
            with_step=False,
            init_kwargs=dict(
                entity='sebastian-garcia-acosta',
                project='marsemseg',
                name='fcn_unet_s5-d16_4x4_512x1024_160k_cityscapes'),
            log_checkpoint=True,
            log_checkpoint_metadata=True,
            num_eval_images=10)
    ])
```

The validation metrics are logged pretty nicely for each class.
![image](https://user-images.githubusercontent.com/46965003/189512624-220c2cfb-0da6-4c44-a47c-895da05e6bac.png)

The hints that pointed me to change the that parameter were these discussions:

- https://github.com/open-mmlab/mmsegmentation/issues/1068
- https://github.com/open-mmlab/mmsegmentation/pull/1603
- https://github.com/open-mmlab/mmcv/pull/1390

I think it is caused because, by default, the validation iterations are not part of regular training steps, so if `with_step = True`, then the metric logs produced by the validation workflow are skipped.



Hope it helps.

**Comment by kimc19** [comment]
Thank you so much for you reply, I'll do some tests with the ``with_step=False`` to see if it works for me.

**Comment by kimc19** [comment]
Thanks for the replys, I add the `with_step=False` parameter in the log_config.
 Also I set `workflow = [('train', 1), ('val', 1)]`  with `samples_per_gpu=1` for the validation loop as it says in https://github.com/open-mmlab/mmsegmentation/blob/master/docs/en/tutorials/customize_datasets.md#data-configuration to get the validation loss.
